st3/lsp_utils/v1.py

```diff
@@ -0,0 +1,120 @@
+
+        def on_response(this, params, request_id):
+            handler(params, lambda result: send_response(request_id, result))
+
+        setattr(self.__plugin, method2attr(method), on_response)
+
+
+class NpmClientHandler(AbstractPlugin):
+    # To be overridden by subclass.
+    package_name = None  # type: Optional[str]
+    server_directory = None  # type: Optional[str]
+    server_binary_path = None  # type: Optional[str]
+    # Internal
+    _package_cache_path = None  # type: Optional[str]
+    _cache_server_path = None  # type: Optional[str]
+
+    @classmethod
+    def setup(cls) -> None:
+        assert cls.package_name
+        assert cls.server_directory
+        assert cls.server_binary_path
+        if cls._package_cache_path:
+            return
+        cls._package_cache_path = os.path.join(sublime.cache_path(), cls.package_name)
+        cls._cache_server_path = os.path.join(cls._package_cache_path, cls.server_directory)
+
+    @classmethod
+    def cleanup(cls) -> None:
+        cls._package_cache_path = None
+        cls._cache_server_path = None
+
+    @classmethod
+    def name(cls) -> str:
+        assert cls.package_name
+        return cls.package_name[len('LSP-'):]
+
+    @classmethod
+    def configuration(cls) -> sublime.Settings:
+        if not cls._package_cache_path:
+            # https://github.com/sublimelsp/LSP/issues/899
+            cls.setup()
+        settings = super().configuration()
+        path = os.path.join(sublime.cache_path(), cls.package_name, cls.server_binary_path)
+        settings.set('command', ['node', path, '--stdio'])
+        return settings
+
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        cache_server_path = os.path.join(cls._package_cache_path, cls.server_directory)
+        if os.path.isdir(cache_server_path):
+            # Server already in cache. Check if version has changed and if so, update.
+            try:
+                src_path = 'Packages/{}/{}/'.format(cls.package_name, cls.server_directory)
+                dst_path = 'Cache/{}/{}/'.format(cls.package_name, cls.server_directory)
+                src_package_json = ResourcePath(src_path, 'package.json').read_text()
+                dst_package_json = ResourcePath(dst_path, 'package.json').read_text()
+                if src_package_json == dst_package_json:
+                    return False
+            except FileNotFoundError:
+                pass
+        return True
+
+    @classmethod
+    def install_or_update(cls) -> None:
+        cache_server_path = os.path.join(cls._package_cache_path, cls.server_directory)
+        src_path = 'Packages/{}/{}/'.format(cls.package_name, cls.server_directory)
+        ResourcePath(src_path).copytree(cache_server_path, exist_ok=True)
+        cmd = ["npm", "install", "--verbose", "--production", "--prefix",
+               cls._cache_server_path, cls._cache_server_path]
+        subprocess.check_call(cmd, shell=sublime.platform() == 'windows')
```

Comment:
Would be good to port over some parts from ServerNpmResource:
 - activity indicator
 - some status messages (including on error)

Reply:
That is something for the base LSP package. Exceptions are already caught in the familiar way by showing an error dialog, same as when your server subprocess can't start.